Teach compiletest to parse arbitrary cfg options

inquisitivecrystal opened the pull request:
This gives compiletest the ability to check for any cfg option that rustc sets by default for certain platforms. For instance, it is now possible to only run a test on targets where `target_has_atomic="8"` is true. 

Resolves #87377.

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3141664086) [(plain)](https://github.com/rust-lang/rust/commit/dc1f5d4d4d9fb65c8bfcf1f5e7ad6374daf963f7/checks/3141664086/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustfix v0.6.0
   Compiling compiletest v0.0.0 (/checkout/src/tools/compiletest)
    Finished release [optimized] target(s) in 20.29s
Check compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:218:10



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:08:55
```

</details>

Urgau commented on src/tools/compiletest/src/main.rs:
```diff
@@ -211,6 +211,13 @@ pub fn parse_config(args: Vec<String>) -> Config {
         // Avoid spawning an external command when we know tidy won't be used.
         false
     };
+    let rustc_cfg_data = Command::new("rustc")
```
You should use the `rustc-path` from the config below instead of the plain `rustc` that you wrote.

inquisitivecrystal replied:
Thank you!

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3141955964) [(plain)](https://github.com/rust-lang/rust/commit/2db8489b6eeef045cb15c811708542ba66909311/checks/3141955964/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
test header::tests::asm_support ... FAILED

failures:

---- header::tests::channel stdout ----
thread 'header::tests::channel' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
---- header::tests::llvm_version stdout ----
---- header::tests::llvm_version stdout ----
thread 'header::tests::llvm_version' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
---- header::tests::aux_build stdout ----
---- header::tests::aux_build stdout ----
thread 'header::tests::aux_build' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
---- header::tests::no_system_llvm stdout ----
error: test failed, to rerun pass '--bin compiletest'
error: test failed, to rerun pass '--bin compiletest'
thread 'header::tests::no_system_llvm' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
---- header::tests::debugger stdout ----
---- header::tests::debugger stdout ----
thread 'header::tests::debugger' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
---- header::tests::ignore_target stdout ----
---- header::tests::ignore_target stdout ----
thread 'header::tests::ignore_target' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
---- header::tests::should_fail stdout ----
---- header::tests::should_fail stdout ----
thread 'header::tests::should_fail' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
---- header::tests::stage stdout ----
---- header::tests::stage stdout ----
thread 'header::tests::stage' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
---- header::tests::sanitizers stdout ----
---- header::tests::sanitizers stdout ----
thread 'header::tests::sanitizers' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
---- header::tests::revisions stdout ----
---- header::tests::revisions stdout ----
thread 'header::tests::revisions' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
---- header::tests::test_duplicate_revisions stdout ----
---- header::tests::test_duplicate_revisions stdout ----
thread 'header::tests::test_duplicate_revisions' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
note: panic did not contain expected string
      panic message: `"called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: \"No such file or directory\" }"`,
 expected substring: `"Duplicate revision: `rpass1` in line ` rpass1 rpass1`"`
---- header::tests::cross_compile stdout ----
thread 'header::tests::cross_compile' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
---- header::tests::only_target stdout ----
---- header::tests::only_target stdout ----
thread 'header::tests::only_target' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10
---- header::tests::asm_support stdout ----
---- header::tests::asm_support stdout ----
thread 'header::tests::asm_support' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/compiletest/src/main.rs:219:10

failures:
    header::tests::asm_support
    header::tests::aux_build
```

</details>

inquisitivecrystal commented on src/tools/compiletest/src/main.rs:
```diff
@@ -211,11 +213,24 @@ pub fn parse_config(args: Vec<String>) -> Config {
         false
     };
+
+    let target_cfg = if cfg!(test) {
```
I'm not sure this is great, but I don't know how else to get around the fact that a `rustc_path` seems not to be provided for test runs. There could be something obvious I'm missing though. I'm pretty tired.

inquisitivecrystal commented on src/tools/compiletest/src/main.rs:
```diff
@@ -211,11 +213,24 @@ pub fn parse_config(args: Vec<String>) -> Config {
         false
     };
+
+    let target_cfg = if cfg!(test) {
+        HashMap::new()
+    } else {
```
I'm not sure this is great, but I don't know how else to get around the fact that a rustc_path seems not to be provided for test runs. There could be something obvious I'm missing though. I'm pretty tired.

Mark-Simulacrum replied:
Hm, it seems likely that should be fixable- maybe a separate PR is best, but it seems like we can pass mostly the same args during tests via an env variable or something...

Can you leave a FIXME comment here perhaps?

inquisitivecrystal replied:
Sure!

inquisitivecrystal commented on src/tools/compiletest/src/util.rs:
```diff
@@ -211,6 +213,55 @@ pub fn logv(config: &Config, s: String) {
+    required_value: Option<&str>,
+) -> bool {
+    let name = name.replace("-", "_");
+    let required_value = required_value.map(str::trim).map(str::to_string);
```
Note to self: change this to one map operation with a closure.

nagisa commented:
I don't see how this handles situations where the test specifies specific flags for rustc, e.g. like this:

```rust
// compile-flags: --target=wasm32-unknown-unknown -Ctarget-feature=+simd128
// cfg-target-feature: simd128
```

should probably work.

inquisitivecrystal commented:
> I don't see how this handles situations where the test specifies specific flags for rustc, e.g. like this:
> 
> ```rust
> // compile-flags: --target=wasm32-unknown-unknown -Ctarget-feature=+simd128
> // cfg-target-feature: simd128
> ```
> 
> should probably work.

It doesn't. I agree that it would be lovely if that were supported. However, it's more complex to implement, because that information isn't presently available to the section of the code that parses configurations. The current version of per target configuration doesn't support that either, so I don't think it makes a lot of sense to block this on adding it. Additionally, there's a relatively easy workaround for that case: just check whether the feature you want is supported on the target you have requested. It's not perfect, but it's a lot more workable than checking whether the feature is supported on every individual target, having to account for the possibility that future targets with different features might be added.

Mark-Simulacrum commented on src/tools/compiletest/src/common.rs:
```diff
@@ -275,6 +276,10 @@ pub struct Config {
 
+    /// A map from target cfg keys to their values. If the key is present with no value,
+    /// the vector will contain the empty string.
+    pub target_cfg: HashMap<String, Vec<String>>,
```
I think it'd be useful to give a short example here (e.g. the has atomic kv pairs).

Mark-Simulacrum commented on src/tools/compiletest/src/header.rs:
```diff
@@ -642,14 +644,42 @@ impl Config {
+        // Matches in full:
+        //   cfg-target-has-atomic: 128
+        //   cfg-target-has-atomic:128
+        //   cfg-target-has-atomic
```
cfg: target-has-atomic=128 rather than this syntax might be a bit better - it matches more closely the syntax used in cfg and similar, which seems positive (and more easily extendable to any or other directives in the future).

inquisitivecrystal replied:
We could I guess? The reason I'm hesitating is that this syntax is standard for compiletest header directives. `=` would require you to remember that this particular directive works differently from all other directives, including other `only` and `ignore` directives, which are otherwise the same. That seems more confusing than breaking with the cfg format. This is also why I've used dashes, rather than underscores, for the examples.

Mark-Simulacrum replied:
None of the other directives are trying to emulate cfg() attributes essentially; they have their own ad-hoc syntaxes. I think in this case it's closer to "this directive matches the stuff you can put inside cfg (with some limitations currently)".

inquisitivecrystal replied:
Oh, wait, I totally misread your comment. Yeah, that... would work fine, actually. I can do that.

Mark-Simulacrum commented on src/tools/compiletest/src/header.rs:
```diff
@@ -642,14 +644,42 @@ impl Config {
+        // Matches up to the second space (exclusive):
+        //   cfg-target-has-atomic: 128 - this requires fancy newfangled atomics
+        //
+        // Matches up to the second colon (exclusive)
```
I'm not following the intent of this case - what is this useful for?

inquisitivecrystal replied:
Comments. There are a bunch of situations where we're using comments on the same line as a directive, using `- comment` or whatever. I didn't want to break those. I could leave a comment here explaining that?

Mark-Simulacrum replied:
Those aren't present on any cfg- lines though, right?

I don't think we should use a colon to terminate lines anyway, that seems "weird". Ultimately I'm not too worried about supporting comments *on the same line* -- the only worry would be if the commit causes us to silently not run a test (which the author may not notice). Not sure how much we can detect mistakes here though... I recall there was some logic added to this effect in Cargo(?) recently, but not 100% sure if that works for us or we can use it...

Mark-Simulacrum commented on src/tools/compiletest/src/util.rs:
```diff
@@ -211,6 +213,55 @@ pub fn logv(config: &Config, s: String) {
 
+/// Parses the output of `rustc --print cfg` into a [`HashMap`] containing
+/// the cfg values for each name. If the name is present with no value,
+/// it is treated as having a value of "".
```
Does this match rustc's handling of no value cfgs? If not, is there a reason for using an empty string instead of an empty vector?

inquisitivecrystal replied:
Well, I thought that rustc is inconsistent. In some cases it has just the value, with no `=`, and in other cases it has the empty string as a value. Looking more closely, there is a semantic difference. The empty string indicates that there could be a value but isn't one, whereas no value indicates that there couldn't be a value. For example, my Mac has `target_abi=""` but also `unix`.

I standardized on the empty string because that was easier to code into `cfg_add`. There's no way to match on the empty string specifically, and no point in adding one, so it's pretty irrelevant for our purposes.  I can change it if you want. I do like the elegance of this code though... :)

Mark-Simulacrum commented on src/tools/compiletest/src/util.rs:
```diff
@@ -211,6 +213,55 @@ pub fn logv(config: &Config, s: String) {
+        values.push(value.to_string());
+    } else {
+        map.insert(name, vec![value.to_string()]);
+    }
```
Fwiw I usually write this sort of thing with the entry API, entry(key).or_default().push(...).

inquisitivecrystal replied:
I can do that. It's a bit cleaner.

Mark-Simulacrum commented on src/tools/compiletest/src/util.rs:
```diff
@@ -211,6 +213,55 @@ pub fn logv(config: &Config, s: String) {
+        (Some(_), None) => true,
+        (Some(values), Some(required_value)) => values.contains(&required_value),
+    }
+}
```
I think if we go with my prior suggestion and use `cfg: foo` as the syntax, we can get rid of all this parsing and just look for a matching line in the rustc output, right?

inquisitivecrystal replied:
We'd have to do prefix based matching, if we wanted to keep the ability to do check for presence of an option without looking at its value. I'm having trouble remembering what the use case was for that... Other than that, we'd have to do some normalization on the rustc output or the header directive, to make them follow the same format while keeping it ergonomic. Do you think it's a good idea to make the change?

Mark-Simulacrum replied:
What normalization are you expecting? I would expect a 1-1 match between the syntax here and in rustc output, I think...?

inquisitivecrystal replied:
The normalization would be removing quotes and spaces. Removing quotes isn't strictly necessary, but makes things a bit easier to use and decreases the chance of errors. Removing spaces is necessary to match `cfg!`, which allows spaces. rustc's output doesn't have any spaces.

nagisa commented:
> Given all that, I don't think we should block this on adding compile-flags support.

I think it is fairly important that writing a test is easy to get right intuitively, and perhaps combining concepts seen in other tests. My view of this specific implementation is that the behaviour is not intuitive at all.

I don't mind this being a separate PR, but from what I can tell, a better integrated implementation would have to redo most of this PR either way. At the very least I want to see a tidy lint that disallows tricky combinations, if we land this feature as-is.

inquisitivecrystal commented:
> > Given all that, I don't think we should block this on adding compile-flags support.
> 
> I think it is fairly important that writing a test (possibly by combining concepts seen in other tests) is easy to get correct. My view of this specific implementation is that the behaviour is not intuitive at all.
> 
> I don't mind this being a separate PR, but from what I can tell, a better integrated implementation would have to redo most of this PR either way. At the very least I want to see a tidy lint that disallows tricky combinations, if we land this feature as-is.

That seems fair. This isn't any worse than the other `only` and `ignore` directives, but one could say that the behavior is confusing there too. When I have a chance, I'll have a go at actually supporting it, and if that seems like too much work I'll add the tidy check. I'll also implement Mark-Simulacrum's comments at the same time.

 Also, you're right that a PR that did that would have to rewrite a fair portion of this: I'd estimate about 50%.

inquisitivecrystal commented:
I think I'm going to temporarily close this. I am working on it again though. Hopefully I'll have progress soon, and I'll reopen then.

inquisitivecrystal commented:
I feel like this is worth reopening for another round of comments. I believe that I've implemented all of the feedback I previously received. I'm sincerely sorry that this took so long. It's been a busy few months, but that's no excuse.

Things that are missing from this implementation are caching of the results and testing. I'd like hear people's thoughts about whether caching is necessary and how to go about it; I'll leave a comment on the relevant area of code. I think testing would likely be advisable, but I'm still working out the best way to go about it. I'm also clarifying some of the relevant details with someone else at the moment.

One other specific thing that I wanted to draw attention to is that I changed the code to refer to the sort of thing that goes after an `ignore-` as a "prop", for property. They were previously called `cfg`s. The problem was that this file uses `cfg` and `config` and `Config` to refer to so many things that I was finding it difficult to reason about which was which. Additionally, although there are some analogies between these properties and `#[cfg(...)]` directives, they aren't all the sort of thing that would go in that kind of directive. Feedback is welcome.

Please feel welcome to point out anything that I've missed, and of course any new comments as well. I'm sure there are places where the code could be improved. CC @Mark-Simulacrum and @nagisa, who reviewed this previously.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4572999157?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/e600e46b60f72d2b3e6d38421a077b38a06c5a90/checks/4572999157/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [assembly] assembly/asm/aarch64-outline-atomics.rs stdout ----

error: compilation failed!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/assembly/asm/aarch64-outline-atomics.rs" "-Zthreads=1" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/asm/aarch64-outline-atomics/aarch64-outline-atomics.s" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-O" "--target" "aarch64-unknown-linux-gnu" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/asm/aarch64-outline-atomics/auxiliary" "--emit=asm"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0463]: can't find crate for `std`
  |
  = note: the `aarch64-unknown-linux-gnu` target may not be installed
  = help: consider downloading the target with `rustup target add aarch64-unknown-linux-gnu`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`

error[E0412]: cannot find type `AtomicI32` in this scope
   |
   |
12 | pub fn compare_exchange(a: &AtomicI32) {

error[E0425]: cannot find value `Relaxed` in this scope
  --> /checkout/src/test/assembly/asm/aarch64-outline-atomics.rs:15:39
   |
   |
15 |     let _ = a.compare_exchange(0, 10, Relaxed, Relaxed);

error[E0425]: cannot find value `Relaxed` in this scope
  --> /checkout/src/test/assembly/asm/aarch64-outline-atomics.rs:15:48
   |
   |
15 |     let _ = a.compare_exchange(0, 10, Relaxed, Relaxed);

error: requires `sized` lang_item

error: aborting due to 5 previous errors
---
test result: FAILED. 117 passed; 1 failed; 18 ignored; 0 measured; 0 filtered out; finished in 0.60s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/assembly" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "assembly" "--mode" "assembly" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:14:50
```

</details>

inquisitivecrystal commented on src/tools/compiletest/src/header.rs:
```diff
@@ -704,12 +736,31 @@ impl Config {
+            .unwrap()
+            .stdout;
+        let cfg_data = String::from_utf8(cfg_data).unwrap();
+        let cfg_data = cfg_data.replace('"', "");
```
The `cfg_data` could presumably be cached somewhere around this point. Doing so could presumably result in some small performance saving for the case where `target == self.target`, which is likely to come up with the most frequency. I'm not sure whether this is worthwhile, or if so, how to go about it.

As an amusing anecdote, I actually fiddled around with a version where all of the targets had their information calculated in advance and it was stored in `Config`. I figured that computing all of that would only take a few extra seconds. It took me a few days to figure out why doing so was causing a one third to one half regression in testing times. It turns out that `Config` gets cloned a fair number of times, and by the end was taking enough memory that it was vastly increasing the duration of forking the threads to execute the tests.

inquisitivecrystal commented:
I added tests!

Mark-Simulacrum commented:
I think I agree with @nagisa that this should respect compile-flags -- or, at least, abort if any such flags are specified so that it can't be combined with compile-flags by accident in a way that leads to misleading behavior. Is that implemented in the current version of the PR?

I'll also note that I feel a little uncertain about the exact shape of these options. It feels like they're an improvement over the manual listing of various targets to ignore, but are a little weak in terms of readability -- `ignore-cfg: ...` doesn't obviously suggest ignore *if* this cfg is true to me. Maybe `ignore-if-cfg` would be better here, not sure.

inquisitivecrystal commented:
> I think I agree with @nagisa that this should respect compile-flags -- or, at least, abort if any such flags are specified so that it can't be combined with compile-flags by accident in a way that leads to misleading behavior. Is that implemented in the current version of the PR?

Well, I thought it was, but... I only made it respect the `--target` compile-flag. Looking back at nagisa's example, that was a mistake. `-Ctarget-feature` is relevant as well, for instance. I'm not sure why someone would set a target feature and then check for it, but it should probably work if they do. I'll change it to just use all of the flags. The target flag will still need special-casing, so this is going to get interesting.

> I'll also note that I feel a little uncertain about the exact shape of these options. It feels like they're an improvement over the manual listing of various targets to ignore, but are a little weak in terms of readability -- `ignore-cfg: ...` doesn't obviously suggest ignore _if_ this cfg is true to me. Maybe `ignore-if-cfg` would be better here, not sure.

I feel uncertain about that too. `ignore-cfg: ...` isn't the clearest. `ignore-if-cfg` seems like a clarity improvement, but still not ideal. Perhaps I should solicit opinions on Zulip? Programmers are terrific at bikeshedding. :)

bors commented:
:umbrella: The latest upstream changes (presumably #94134) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
@inquisitivecrystal any updates on this?

inquisitivecrystal commented:
> @inquisitivecrystal any updates on this?

I apologize for the delay. I know this PR has been hanging around for ages at this point. I'm still planning on finishing everything up sooner or later, but it may be a month or so before I have time to work on it again, as I'm extremely busy IRL. Unfortunately, the "proper" fix is so complicated that I don't know how to implement it cleanly (and the difference wouldn't be relevant all that often, in any case). When I do implement this, I'm planning to go with a simpler implementation (using tidy to reject confusing combinations, rather than trying to support them). 

In the meantime, I don't know how to handle the PR. We can close it so WG-triage doesn't have to deal with keeping track of it or leave it open if that's more convenient. I'm even happy to let someone else take over, if there's someone who wants to work on that. In any case, I once again apologize for how long it's taken me to handle this.

Dylan-DPC commented:
Nothing to apologise about. I am going to close it for now, but feel free to reoopen it if nobody has got to it yet.
